### PR TITLE
chromium: Build with VA-API but disable it by default

### PR DESCRIPTION
**Pull request by primeos**
This makes it possible to enable VA-API without having to rebuild
Chromium.

Co-Authored-By: ElXreno <elxreno@gmail.com>
Co-Authored-By: misuzu <bakalolka@gmail.com>

###### Motivation for this change

Alternative implementation of #85076. Solves the VA-API part of #82443.

Advantages:
- Doesn't require `--ignore-gpu-blacklist`, i.e. VA-API can be configured independently of other GPU features
- Will explicitly break all existing VA-API setups (vs. silently). Bonus: We could add instructions via a `throw` statement.
- Doesn't interfere with `commandLineArgs`. I.e. we don't accidentally enable VA-API if that argument is set.

Disadvantages:
- More complex!

TODOs:
- [x] Add a `throw` statement for `useVaapi`?
- [x] Check the Ozone build (WIP). At least on M80 the build did fail (I wrote a possible fix, but for now it seems best to disable VA-API for Ozone).
- [x] Add some documentation
- [x] Verify the build on AArch64
- [x] Update the NixOS wiki after this PR (documentation)
  - https://nixos.wiki/wiki/Chromium#Enable_GPU_accelerated_video_decoding_.28VA-API.29
  - https://nixos.wiki/wiki/Accelerated_Video_Playback#Chromium

Tests:
- [x] On a working VA-API setup
- [x] Without any VA-API drivers
- [x] With XWayland (expected: VA-API doesn't work, depending on the driver)
  - `enableVaapi = false`: Works as expected
    - Only drawback: Additionally prints `vaInitialize failed: unknown libva error`
    - But: `chrome://gpu` now even shows the correct `Video Decode` information: `Unavailable` -> `Software only. Hardware acceleration disabled`
  - `enableVaapi = true`: Wrong VA information, but ok
    - The same as before, but `chrome://gpu` now incorrectly shows `Video Decode: Hardware accelerated` (despite the libva error and VA not actually working)
- [x] Optional: In a VM (`virtio`)
- [x] Invalid `LIBVA_DRIVER_NAME` value: Fallback to SW rendering (only wrong VA information shown, like in the second XWayland case)
- [x] VA-API on `radeon` almost works now (at least on my old HD 6950) \o/
  - Was broken before, probably fixed in the meantime by a Mesa update
  - Update: There are some minor visual glitches left at the bottom of the screen but it's watchable (vs. completely visually broken as before)
- Tested drivers: `i965`, `iHD`, and `r600` (some visual glitches with VA-API, but ok)

Note: We could also add `chromium-vaapi` but I'd like to avoid that for now (as we're still experimenting with it).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/browsers/chromium/default.nix [hunk 1]
```diff
@@ -127,6 +127,10 @@ let
         cp -a ${widevineCdm}/WidevineCdm $out/libexec/chromium/
       ''
     else browser;
+
+  optionalVaapiFlags = lib.optionalString
+    (!enableVaapi)
```
Comment (offlinehacker) [inline pkgs/applications/networking/browsers/chromium/default.nix:132]: Why is this needed? Will not chrome disable by default?

**Comment by offlinehacker** [comment]
For some reason I tought this was already enabled. I think we should build with va-api by default, it's pretty common feature.

File: pkgs/applications/networking/browsers/chromium/default.nix (hunk 1 above)
Comment (primeos) [reply pkgs/applications/networking/browsers/chromium/default.nix]: No, we changed that behaviour (see the linked PR and issue for the discussion).

**Comment by primeos** [comment]
I added some minor improvements but it's basically a refactoring except for the Ozone "fix" and added deprecation note via `throw` (but the derivations for `chromium` and `chromium.override { enableVaapi = true; }` remain unchanged).

**Comment by primeos** [comment]
@thefloweringash nearly forgot to ask: Could you test this on AArch64 to see if the build still works or if we have to disable VA-API there?

**Comment by thefloweringash** [comment]
I'm not sure what to make of this.

On Aarch64 it does build and run, but the GPU thread crashes during startup and prints a [backtrace including VADisplayState::Initialize](https://gist.github.com/thefloweringash/13323384d9b7177ca3e9bc8d3a842636). Inspecting `chrome://gpu` shows the following warning:
> GPU process was unable to boot: GPU process crashed too many times with SwiftShader.
Disabled Features: all

Software webgl is no longer available. This happens even with the default chromium running with `--disable-accelerated-video-decode --disable-accelerated-video-encode`.

This could be a symptom of my particular graphics stack. I only have one graphical system to test this on at the moment and I'm using the panfrost driver which IIUC is still under active development.

**Comment by primeos** [comment]
@thefloweringash huge thanks for your testing, that's very helpful! As for AArch64 I decided to disable VA-API for now as I'm not sure if we could get this to work properly across all platforms or if we even package VA-API drivers for GPUs on ARM systems.

If you have time for it: Could you test the new version of this PR again (which would also cover #85360) to verify that the regressions are gone. The only difference for AArch64 should now be the additional `--disable-accelerated-video-decode --disable-accelerated-video-encode` flags (could be avoided as well but then we might risk enabling VA-API at runtime if we should build it for AArch64 in the future).

**Comment by primeos** [comment]
Note: I'll try to run some more tests on `x86_64`, seems like this change is more complicated than anticipated. I also realized that I did only test this on systems where I tested VA-API before and it's properly supported. I'll try to test the following cases as well:
- Without any VA-API drivers
- With XWayland (doesn't work, depending on the driver)
- Optional: In a VM

**Comment by thefloweringash** [comment]
Tested on aarch64: some browsing, youtube and software webgl are working. LGTM.

**Comment by primeos** [comment]
My testing on `x86_64` didn't show any important regressions, let's hope this'll work well with any GPU+driver combination (at least for the default case with `enableVaapi = false`). Also this PR should at least avoid the regressions from the last PR since we disable VA-API at runtime by default for now.

I'll merge it now :tada:

Note: The last force-push was just to clean up the history (rebase+squashing - no other changes).

**Comment by offlinehacker** [comment]
`enableVaapi`does not seem to be passed to `mkChromium`

**Comment by primeos** [comment]
@offlinehacker that's intentional, it only affects the wrapper ;)
